Change User-Extended App user group API mechanism

anagperal opened the pull request:
### :pushpin: References

-   **Issue:** #8699r8t7c Closes [Change User-Extended App user group API mechanism](https://app.clickup.com/t/8699r8t7c)

- **OLD** DHIS2 Documentation: https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-241/metadata.html?h=develop#webapi_collections_adding_removing_objects_single_request
- **NEW** DHIS2 Documentation: https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-241/metadata.html?h=develop#webapi_partial_updates

### :memo: Implementation

- [x] Replace old API mechanism and use patch method instead when saving userGroups.  Now the lastUpdated field of the user group changes, ant the entry will appear in the audit log
- [x] Fix tests: update Jest config to allow axios and other ES modules to be transformed

### :video_camera: Screenshots/Screen capture

[Screencast from 2025-08-13 15-02-36.webm](https://github.com/user-attachments/assets/20f2ea2b-6df4-48d4-b03b-359b22e83f51)

tested using https://play.im.dhis2.org/stable-2-40-9 :

[Screencast from 2025-09-12 09-59-22.webm](https://github.com/user-attachments/assets/dbbd5530-758c-4421-ac49-973eed637786)


### :fire: Testing

It needs new version of d2-api: https://github.com/EyeSeeTea/d2-api/pull/176

To test it:

Go to that branch of d2-api:
```shell
yarn install
yarn build
cd build
yarn link
```

On this app:

```shell
yarn link "@eyeseetea/d2-api"
```

anagperal committed:
Replace old API mechanism and use patch method instead when saving userGroups

ifoche commented:
Task linked: [CU-8699r8t7c Change User-Extended App user group API mechanism](https://app.clickup.com/t/8699r8t7c)

anagperal committed:
Fix tests: update Jest config to allow axios and other ES modules to be transformed

anagperal committed:
Add types to map to fix git actions

ifoche reviewed (changes requested):
Hey @anagperal . As you know, I'm recently starting by doing a quick codex review before doing the functional one. Here's Codex 5 cents:

- Potential Issue: The updateUserGroups implementation incorrectly computes group membership changes.
It only patches groups that did not previously exist and, when removing members, it attempts to remove all users from a group instead of just the affected ones. This can lead to users not being added to existing groups and to unintended removal of all current members.

Suggested task
Fix group-membership diff logic in updateUserGroups


- Potential Issue: User deletions still rely on the legacy metadata POST endpoint with importStrategy=DELETE, which is deprecated in recent DHIS2 versions. A modern API call should be used instead to remove users.

Suggested task
Replace deprecated user deletion endpoint


- Potential Issue: updateUserGroups logs group IDs but omits user-level details, which complicates debugging when partial updates fail or produce unexpected results.

Suggested task
Improve logging for group patch operations


Could you address those or justify if there's no need for a particular one?

Thank you!

anagperal committed:
Fix potential issue when updating user groups

anagperal committed:
Improve log

tokland commented on i18n/es.po:
```diff
@@ -557,10 +569,6 @@ msgstr "Abrir en Administración de usuarios"
-msgid "and_n_more"
-msgstr ""
-
 #, fuzzy
```
Now that we are at it, let's fix all the fuzzy references in the file

anagperal commented:
> Hey @anagperal . As you know, I'm recently starting by doing a quick codex review before doing the functional one. Here's Codex 5 cents:
> 
> * Potential Issue: The updateUserGroups implementation incorrectly computes group membership changes.
>   It only patches groups that did not previously exist and, when removing members, it attempts to remove all users from a group instead of just the affected ones. This can lead to users not being added to existing groups and to unintended removal of all current members.
> 
> Suggested task Fix group-membership diff logic in updateUserGroups
> 
> * Potential Issue: User deletions still rely on the legacy metadata POST endpoint with importStrategy=DELETE, which is deprecated in recent DHIS2 versions. A modern API call should be used instead to remove users.
> 
> Suggested task Replace deprecated user deletion endpoint
> 
> * Potential Issue: updateUserGroups logs group IDs but omits user-level details, which complicates debugging when partial updates fail or produce unexpected results.
> 
> Suggested task Improve logging for group patch operations
> 
> Could you address those or justify if there's no need for a particular one?
> 
> Thank you!

Thanks @ifoche !

These tasks are not related to the current one, but since one of them was a bug, I fixed it in this PR. Good catch of Codex!

1. Suggested task Fix group-membership diff logic in updateUserGroups: I have analysed the feature and there was a bug when for example choosing two users, user1 with userGroups A and B and user2 with userGroup A, and we try to add both userGroups A,B to both users. In this case, due to a bug in the code userGroup B was never going to be added in user2. Now it it's fixed. Also Codex says that we can have an unintended removal of all current members, but this cannot happend as when you select more that one user it only allows you to delete user groups that are common in all users.
2. Suggested task Replace deprecated user deletion endpoint: I have analysed the documentation of DHIS2 and I cannot find any information about the deprecation of importStrategy=DELETE
3. Suggested task Improve logging for group patch operations: I have improve the log

tokland commented on src/data/utils.ts:
```diff
@@ -33,3 +33,17 @@ export function buildUserWithoutPassword(users: ApiUser[]) {
+
+        const diffUserIds = sourceUserGroup.filter(user => !userIdsInRefGroup.includes(user.id)).map(({ id }) => id);
+
+        return { id: groupId, usersIds: [...new Set(diffUserIds)] };
```
is the intent equivalent to `_.uniq`?

anagperal replied:
Yes, I have changed it

tokland reviewed (commented):
Some minor in-line comments:

About the d2-api link: typically we release a beta so it's easy to test the feature branch in the app.

When trying to do so, found a problem. https://github.com/EyeSeeTea/d2-api/pull/176 points to a branch with only 2.40 schemas. But this app links to 2.36 schemas. What's the idea? We either update the app to 2.40 references or we'll need another branch in d2-api.

anagperal committed:
Use unique instead of new Set and delete fuzzy in translations

anagperal commented:
> Some minor in-line comments:
> 
> About the d2-api link: typically we release a beta so it's easy to test the feature branch in the app.
> 
> When trying to do so, found a problem. [EyeSeeTea/d2-api#176](https://github.com/EyeSeeTea/d2-api/pull/176) points to a branch with only 2.40 schemas. But this app links to 2.36 schemas. What's the idea? We either update the app to 2.40 references or we'll need another branch in d2-api.

thanks @tokland ! I didn't realise about this because I have tested with an build I have in d2-api where I had 2.36 as I didn't run yarn clean before yarn build. 

I'll let @ifoche decide about it, but since we have this task in TO-DO https://app.clickup.com/t/869a4umgc maybe if there is no urgent need to complete this task, we can wait until 2.41 schemas are done, if not we'll need the branch with the 2.36 schemas

tokland reviewed (approved):
Approved, code-wise, let me know when you need a beta to be released after deciding on the d2-api version to use.

ifoche commented:
@anagperal just to be sure that we can merge this PR. The situation is:
- I requested some changes, that I understand you addressed, but you didn't request for a new review, so the status of the PR is still showing up as me requesting changes
- After my request of changes, Arnau requested some additional ones that I understand you addressed them, however, Arnau finished by asking you whether you would need a new d2-api release... is that done? if so... shouldn't we need to change this code pointing to that new release?

Thanks

anagperal committed:
Point to 2.41 schemas

anagperal commented:
> @anagperal just to be sure that we can merge this PR. The situation is:
> 
> * I requested some changes, that I understand you addressed, but you didn't request for a new review, so the status of the PR is still showing up as me requesting changes
> * After my request of changes, Arnau requested some additional ones that I understand you addressed them, however, Arnau finished by asking you whether you would need a new d2-api release... is that done? if so... shouldn't we need to change this code pointing to that new release?
> 
> Thanks

@ifoche Yes, all comments has been addressed. For having PATCH of d2-api we have two options (cc @tokland ):
1. Wait until this PR is done  https://github.com/EyeSeeTea/user-extended-app/pull/323  (now in progress but I have to change to more urgent tasks)
2. point to another branch in d2-api with 2.36 schemas generated

tokland commented:
> @ifoche Yes, all comments has been addressed. For having PATCH of d2-api we have two options (cc @tokland ):
> 
>     1. Wait until this PR is done  [Point to 2.41 schemas #323](https://github.com/EyeSeeTea/user-extended-app/pull/323)  (now in progress but I have to change to more urgent tasks)
> 
>     2. point to another branch in d2-api with 2.36 schemas generated

Up to @ifoche, but it'd be nice to upgrade to the 2.41 schemas d2-api now as a way to validate the other PR.

anagperal committed:
Fix acces as Ref instead of same as AccessPermissions type

anagperal commented:
@ifoche @tokland PR https://github.com/EyeSeeTea/user-extended-app/pull/323 ready to review, so it's better to wait to v2.41

anagperal committed:
import @eyeseetea/d2-api from d2-api types

anagperal committed:
Remove @ts-ignore

anagperal committed:
Add same timestamp to teiId and enrollmentId

anagperal committed:
Rename getD2APiFromInstance to getD2ApiFromInstance

adrianq committed:
Merge pull request #323 from EyeSeeTea/feature/update-d2-api-schema-41

Point to 2.41 schemas

anagperal committed:
Merge development and solve conflicts

anagperal committed:
Update translations

anagperal committed:
Merge development and solve conflicts

anagperal committed:
Update @eyeseetea/d2-api version and fix tests add @eyeseetea/d2-api to jest transformIgnorePatterns

anagperal committed:
Fix translations

anagperal committed:
Fix translations

ifoche commented:
Hi @anagperal I add it just as a comment, as I don't think the things Codex mentions are really very important ones. But I prefer you to evaluate them:

Here’s what I checked and what’s addressed vs. still pending.

  **Group Diff Logic**

  - Addressed: updateUserGroups now computes per-group diffs correctly via getDiffUserIdsByGroup and issues JSON Patch ops per user.
  - Adds users with op: "add", path: "/users/-", value: { id } and removes with op: "remove-by-id", path: "/users", id, so it no longer wipes entire groups.
  - Unit tests exist for the diff behavior: see src/data/repositories/__tests__/getDiffUserIdsByGroup.spec.ts.

  **User Deletion Endpoint**

  - Not addressed: remove(ids) still uses the legacy metadata import with importStrategy: "DELETE":
      - src/data/repositories/UserD2ApiRepository.ts lines ~71–106
      - This should be replaced with per-user DELETE /api/users/{id} (or the modern bulk delete if your DHIS2 version supports it).

  **Logging Improvements**

  - Partially addressed: updateUserGroups logs only group IDs added/removed:
      - logger.log({ groupsIdsToAdd, groupsIdsToDelete })
  - Missing: user-level details. Recommended to log, per group, the arrays of user IDs being added/removed, e.g.:
      - logger.log({ add: [{ groupId, users: [...] }], remove: [{ groupId, users: [...] }] })
      - Optionally, log per-group patch results or failures for easier debugging.

  **Minor Type Clarity (optional)**

  - D2UserGroupByKey is typed as Record<Id, NamedRef[]> but buildUsersByGroupId supplies ApiUser[]. This is structurally fine (users have id/name), but renaming to something like Record<Id, Array<{ id: Id }>> would
  better reflect usage.

  **Verdict**

  - Group membership diff/removal bug: fixed.
  - Deprecated user deletion endpoint: still pending.
  - Logging granularity for group patches: still pending.

  Want me to:
  - Replace the deletion logic with DELETE /api/users/{id} using Future.parallel + chunkRequest?
  - Enhance logging to include per-group, per-user details?

anagperal committed:
Improve logs

anagperal committed:
Add TODO to change deprecated POST using importStrategy: DELETE

anagperal committed:
Truly return the type of the function buildUsersByGroupId

anagperal commented:
> Hi @anagperal I add it just as a comment, as I don't think the things Codex mentions are really very important ones. But I prefer you to evaluate them:
> 
> Here’s what I checked and what’s addressed vs. still pending.
> 
> **Group Diff Logic**
> 
> * Addressed: updateUserGroups now computes per-group diffs correctly via getDiffUserIdsByGroup and issues JSON Patch ops per user.
> * Adds users with op: "add", path: "/users/-", value: { id } and removes with op: "remove-by-id", path: "/users", id, so it no longer wipes entire groups.
> * Unit tests exist for the diff behavior: see src/data/repositories/**tests**/getDiffUserIdsByGroup.spec.ts.
> 
> **User Deletion Endpoint**
> 
> * Not addressed: remove(ids) still uses the legacy metadata import with importStrategy: "DELETE":
>   
>   * src/data/repositories/UserD2ApiRepository.ts lines ~71–106
>   * This should be replaced with per-user DELETE /api/users/{id} (or the modern bulk delete if your DHIS2 version supports it).
> 
> **Logging Improvements**
> 
> * Partially addressed: updateUserGroups logs only group IDs added/removed:
>   
>   * logger.log({ groupsIdsToAdd, groupsIdsToDelete })
> * Missing: user-level details. Recommended to log, per group, the arrays of user IDs being added/removed, e.g.:
>   
>   * logger.log({ add: [{ groupId, users: [...] }], remove: [{ groupId, users: [...] }] })
>   * Optionally, log per-group patch results or failures for easier debugging.
> 
> **Minor Type Clarity (optional)**
> 
> * D2UserGroupByKey is typed as Record<Id, NamedRef[]> but buildUsersByGroupId supplies ApiUser[]. This is structurally fine (users have id/name), but renaming to something like Record<Id, Array<{ id: Id }>> would
>   better reflect usage.
> 
> **Verdict**
> 
> * Group membership diff/removal bug: fixed.
> * Deprecated user deletion endpoint: still pending.
> * Logging granularity for group patches: still pending.
> 
> Want me to:
> 
> * Replace the deletion logic with DELETE /api/users/{id} using Future.parallel + chunkRequest?
> * Enhance logging to include per-group, per-user details?

thanks @ifoche ! 

- User Deletion Endpoint: This will be a new task. I have added a comment with a TODO in order to change from POST metadata using  importStrategy: "DELETE" to  DELETE /api/users/{id}
- [x] Logging Improvements: I have improved the logs logging per group too
- [x] Minor Type Clarity (optional): Improve buildUsersByGroupId function to truly return the type of the function buildUsersByGroupId

anagperal committed:
Fix d2-api version

anagperal commented:
@ifoche check the replicate feature again, now should work:

[Screencast from 2025-09-18 08-40-24.webm](https://github.com/user-attachments/assets/c6b7c31b-c398-4a2a-99b7-9a4641d958cd)


LOGGER error is normal if there is no configuration in the datastore:

<img width="1149" height="326" alt="Screenshot from 2025-09-18 08-46-25" src="https://github.com/user-attachments/assets/b25a7ef4-818f-448b-ba54-dbad53ff6ec7" />
<img width="1149" height="326" alt="Screenshot from 2025-09-18 08-46-22" src="https://github.com/user-attachments/assets/8587a386-6f5d-4baf-99d5-e0d129224e65" />

ifoche reviewed (changes requested):
sorry @anagperal. I now see the calls using the PATCH method but they keep failing for me... could you double check?

<img width="2451" height="866" alt="Image" src="https://github.com/user-attachments/assets/e10b8e6c-5a31-42eb-84ae-e2b31eaa00c3" />

anagperal committed:
Merge development and solve conflicts

MatiasArriola committed:
fix: make userGroups and saveLocales calls sequential to avoid deadlocks

This happened in specific environments only.
These requests might end up locking each other (update userinfo locks + userinfo locks from FK update usersetting)

ifoche committed:
Merge pull request #334 from EyeSeeTea/fix/deadlock-on-replicate

fix: reduce concurrency to avoid deadlocks

ifoche commented:
@anagperal could you please merge this one with development?

anagperal committed:
Merge development and solve conflicts

anagperal committed:
Merge development and solve conflicts

anagperal commented:
> @anagperal could you please merge this one with development?

@ifoche  done!

ifoche commented:
I'm very sorry @anagperal to write this. I can even feel your pain when you will read this. But unfortunately we're still not there. A simple operation trying to enable the user 1f299160-9697-410c-bc12-54ca55b61186 produces a dramatic failure... could you please check it?
<img width="2308" height="914" alt="2025-11-12-221945_hyprshot" src="https://github.com/user-attachments/assets/58e72766-e490-4a32-96c8-fdbc6efa9ea2" />

anagperal committed:
Fix Either getOrThrow when isValidationErrorArray

anagperal committed:
Fix existing user in update and in save user status change from new to create existed

anagperal committed:
Remove unnecessary owner in fields

anagperal commented:
> I'm very sorry @anagperal to write this. I can even feel your pain when you will read this. But unfortunately we're still not there. A simple operation trying to enable the user 1f299160-9697-410c-bc12-54ca55b61186 produces a dramatic failure... could you please check it? <img alt="2025-11-12-221945_hyprshot" width="2000" height="914" src="https://private-user-images.githubusercontent.com/10931645/513550928-58e72766-e490-4a32-96c8-fdbc6efa9ea2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjMwNTE0NjIsIm5iZiI6MTc2MzA1MTE2MiwicGF0aCI6Ii8xMDkzMTY0NS81MTM1NTA5MjgtNThlNzI3NjYtZTQ5MC00YTMyLTk2YzgtZmRiYzZlZmE5ZWEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTEzVDE2MjYwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzZjMyNGI5NTc3MjRhYjVlNzcyMjYwN2VjZDZlMjQzNDBlMDI1NzQ0YjYzYTYxNDM4ZjgyYTE1NzZjZGIyZjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dZpxwvvE08zQdclaQPgXuCcllykQeejr--dJv0sVLy0">

@ifoche I have fix this error and another when assing search/data view/data capture to OUs. Also I found another extrange behavior explained here https://app.clickup.com/t/8699r8t7c?comment=90120169727059, it's fixed too. I'll add to @xurxodev as a PR reviewer just in case he can review the last commits as it is related to this PR https://github.com/EyeSeeTea/user-extended-app/pull/330

anagperal committed:
Create change specific change status methods and use them in use cases

anagperal committed:
Instead of owner add fields

xurxodev reviewed (approved):
thanks @anagperal 

After our live review of commits in conflict with my PR I appoved it

ifoche commented:
@anagperal sorry but... still not yet there. I think this PR is cursed. Please see my screenshot when I try to filter by groups... no group shows up:
<img width="1208" height="864" alt="2025-11-19-191500_hyprshot" src="https://github.com/user-attachments/assets/54a29377-6200-4bd8-a179-2840cec5d78a" />

Could you please check it?

anagperal commented:
> @anagperal sorry but... still not yet there. I think this PR is cursed. Please see my screenshot when I try to filter by groups... no group shows up: <img alt="2025-11-19-191500_hyprshot" width="1208" height="864" src="https://private-user-images.githubusercontent.com/10931645/516437138-54a29377-6200-4bd8-a179-2840cec5d78a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM2MjQxMzYsIm5iZiI6MTc2MzYyMzgzNiwicGF0aCI6Ii8xMDkzMTY0NS81MTY0MzcxMzgtNTRhMjkzNzctNjIwMC00YmQ4LWExNzktMjg0MGNlYzVkNzhhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTIwVDA3MzAzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgzYzZhZmQxZmJmNThlZGY4MzVmYzk0Y2I1NGM0MjQwZWRiNjRlNGU1OGEyYTY3NTIyODExNDUxMDY3YmRmYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.HQA-qMYDbEU6hRZt3XnXmn1rBH5M2pAnGDyJ9lznwSI">
> 
> Could you please check it?

thanks @ifoche . Yes, it's definitely cursed because I tested everything in general and I remember filtering by group as well, but now suddenly both the request to userRoles and userGroups give a CORS error... I'll take a look.

<img width="1866" height="1009" alt="image" src="https://github.com/user-attachments/assets/cb3dbc00-faa6-4ce9-8f93-32d38f7c5bc2" />

anagperal committed:
Use d2-api instead of d2/lib/d2 for filters: userGroups, userRoles and Org units

anagperal commented:
@ifoche In the filters that load user roles, user groups, and organisation units, the requests were implemented using d2/lib/d2 (d2.models...) instead of our d2-api and this caused intermittent CORS errors in local development.

The reason is that d2.models does not use the local baseUrl we configure, it uses the absolute apiEndpoint URLs returned by the DHIS2 /api/schemas endpoint:

[CORS.webm](https://github.com/user-attachments/assets/9cb1c9ac-cadb-42d3-99a4-e13876bdcb2b)

Now it uses d2-api instead of d2/lib/d2, so the requests are using configured local baseUrl so we don't have CORS errors:

[solved_CORS.webm](https://github.com/user-attachments/assets/df4a18ad-f744-47c2-98ed-e8d7663106b6)

I’m also a bit confused about why this didn’t happen earlier, or maybe we didn't notice. These requests should have always caused CORS problems when running on localhost and this issue is not related to the change to using PATCH. It’s also possible that I’m missing something here, since this is my first time developing in this project. I’ll ask about this behavior in the next tech decisions meeting to check.

I'll ask @tokland to check again this last commit as it touches parts of the legacy code

tokland commented on src/data/repositories/UserD2ApiRepository.ts:
```diff
@@ -95,6 +95,7 @@ export class UserD2ApiRepository implements UserRepository {
 
     remove(ids: Id[]): FutureData<Stats> {
         return chunkRequest(ids, userIds => {
+            // TODO: This should be replaced with per-user DELETE /api/users/{id} or the bulk delete if DHIS2 version supports it
```
Here we may add the "why" it should it be replaced. And it says "if bulk supported", but that's what we are doing, right?

tokland commented on src/data/repositories/UserD2ApiRepository.ts:
```diff
@@ -469,52 +470,64 @@ export class UserD2ApiRepository implements UserRepository {
+                logger.log({
+                    action: action,
+                    groupId: group.id,
+                    usersIds: group.usersIds,
```
The orthodox here is "userIds", two plurals sound weird. Change only if it we don't have already data saved, then it's a good time to do it.

tokland commented on src/domain/entities/Either.ts:
```diff
@@ -56,6 +58,10 @@ export class Either<Error, Data> {
                 if (error instanceof Error) {
                     throw error;
                 } else if (Array.isArray(error)) {
+                    if (isValidationErrorArray(error)) {
```
The handling of special error messages here is a bit unusual. Since Either is a generic data structure, the parent should typically manage the specifics. However, because this method is invoked from many places, let’s keep it as‑is, but we may add a comment explaining what we’re doing and why.

tokland commented on src/domain/entities/User.ts:
```diff
@@ -18,16 +18,23 @@ export class User extends Struct<UserProps>() {
+    // TODO: remove generic update method and use specific methods only
     update(props: Partial<UserProps>): Either<ValidationError<User>[], User> {
-        return User.validateAndCreateUser({ ...this, ...props }, false);
+        return User.validateAndCreateUser({ ...this, ...props }, true);
```
[best practices] use options object to contain boolean arguments , so we have a naming when calling it.

tokland commented on src/domain/errors/Errors.ts:
```diff
@@ -3,3 +3,22 @@ export type ValidationError<T> = {
 };
+
+export function isValidationErrorArray<T>(value: unknown): value is ValidationError<T>[] {
+    return (
```
Ideally, in a point of domain where we have full control we should need to handle unknowns and runtime checks, but again, this is related to the Either method called from everywhere, so we can keep it as-is.

tokland commented on src/domain/usecases/SaveUserOrgUnitUseCase.ts:
```diff
@@ -15,7 +14,7 @@ export class SaveUserOrgUnitUseCase {
             const usersToSave = this.applyOrgUnitsToUsers(options);
             return this.saveUsers(usersToSave);
         } catch (error) {
```
The catch was already there, but a reader may wonder why the try/catch if we are working with futures. We may add a comment if we really need it.

tokland commented on src/domain/usecases/SaveUserOrgUnitUseCase.ts:
```diff
@@ -15,7 +14,7 @@ export class SaveUserOrgUnitUseCase {
             return this.saveUsers(usersToSave);
         } catch (error) {
-            return Future.error(i18n.t(`${(error as Error).message}`));
+            return Future.error(`${(error as Error).message}`);
```
That's the problem of try/catch, we need ugly code. Are we sure we will get an Error here?

Best practice: a centralized getErrorMessage(err: unknown): string, for all the places were we do that snippet.

tokland commented on src/domain/usecases/SaveUserStatusUseCase.ts:
```diff
@@ -10,11 +9,11 @@ export class SaveUserStatusUseCase implements UseCase {
             });
             return this.userRepository.save(usersToUpdate).toVoid();
         } catch (error) {
-            return Future.error(i18n.t(`${(error as Error).message}`));
```
Same, I guess the catch is needed because of the getOrThrow. It's old code, we can leave it as-is, but we can keep it in mind for future code.

tokland commented on src/legacy/components/ImportTable.component.js:
```diff
@@ -103,6 +103,7 @@ const styles = {
 };
 
+// NOTICE: component not used anywhere
 class ImportTableOld extends React.Component {
```
Can we then remove it, then?

anagperal replied:
In new PR: remove legacy code + change d2 library by d2-api

tokland commented on src/utils/tests.tsx:
```diff
@@ -8,13 +8,49 @@ import { Instance } from "../data/entities/Instance";
         userRoles: [],
+        organisationUnits: [],
+        dataViewOrganisationUnits: [],
+        searchOrganisationsUnits: [],
```
`searchOrganisationsUnits` looks off, two plurals, is that a real attribute? I think latest DHIS2 use `teiSearchOrganisationUnits`, can you confirm?

anagperal replied:
yes the user from dhis2 has `teiSearchOrganisationUnits`, but checking the code it seems that the User of the domain has `searchOrganisationsUnits` (UserProps), so I just delete the // @ts-ignore TODO by adding the rest of the UserProps

tokland commented on src/data/repositories/UserD2ApiRepository.ts:
```diff
@@ -526,27 +539,36 @@ export class UserD2ApiRepository implements UserRepository {
+                  }
+        );
+        return apiToFuture(this.api.models.userGroups.patch(userGroup.id, patchOperations)).flatMap(d2Response => {
+            if (d2Response.errorReports?.length !== 0) {
```
If `d2Response.errorReports is undefined` then `d2Response.errorReports?.length` -> undefined and undefined !== 0 -> it enters the error branch, but there were no errors.

anagperal committed:
Improve user method, add explanatory comments and fix buildGroupsToSave if

ifoche commented:
@anagperal I still get these 500 errors re-enabling the users
<img width="1530" height="906" alt="2025-12-21-214943_hyprshot" src="https://github.com/user-attachments/assets/47b352b5-926d-40af-9d70-f48dc4f1da4a" />

But that's more or less what we had discussed last time on this. If I upload the app as a zip, then the frequence of the error is lower and when it happens you repeat the enabling operation and then works. I'm not fully sure of this, but approving so we can go ahead with @idelcano tests and see if this is pointed out in his review. 

Thank you!

ifoche reviewed (approved):
thanks @anagperal
